### ISSUE TITLE
Webpack 4 chunkhash can vary between builds

### ISSUE THREAD

**Issue by mikeyoon** [webpack/webpack#7179]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
I'm finding that the chunkhash varies between builds in v4. Using the diff utility between the the different files shows no differences. Attempting to narrow the codebase down reduces the frequency of different hashes.

**If the current behavior is a bug, please provide the steps to reproduce.**
I've been unsuccessful in creating a small test case as it's just too inconsistent and would take such an enormous amount of time for me to figure out by slicing and dicing our codebase.

I'm happy to help as much as I can to find a repro, but I'd need some information about diagnostics I can turn on to figure out what's happening.

**What is the expected behavior?**
The chunkhash should be consistent, as does in version 3.

**If this is a feature request, what is motivation or use case for changing the behavior?**

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**

Webpack 4.6.0, Node 8.9.4, OS X 10.13.4

Here's our webpack.config.js

```js
const path = require('path');
const webpack = require('webpack');
const ProgressBarPlugin = require('progress-bar-webpack-plugin');
const childProcess = require('child_process');
const UglifyJsPlugin = require('uglifyjs-webpack-plugin')
const AssetsPlugin = require('assets-webpack-plugin')
const CleanWebpackPlugin = require('clean-webpack-plugin')
const WebpackBuildNotifierPlugin = require('webpack-build-notifier');

const root = path.resolve(__dirname, './src');
const webappPath = path.resolve(__dirname, '..');
const deployPath = path.resolve(webappPath, './public/assets');

const dev = !process.env.NODE_ENV;
const ci = process.env.NODE_ENV === 'ci';
const prod = process.env.NODE_ENV === 'production';

if (prod) {
  console.log('PRODUCTION MODE ENABLED');
}

// don't hash the filename unless we're building for prod or staging. Hashed filenames interfere with karma tests.
const outputFilename = dev || ci ? '[name].bundle.js' : '[name].[chunkhash].bundle.js';
const excludedMapSources = dev ? [/vendor.*bundle.js/, /manifest.*bundle.js/] : [/manifest.*bundle.js/];

function getPlugins() {
  let plugins = [
    new CleanWebpackPlugin(['public/assets/*'], {
      allowExternal: true,
      root: webappPath
    }),

    new webpack.SourceMapDevToolPlugin({
      filename: "[file].map",
      exclude: excludedMapSources
    }),

    new AssetsPlugin({
      includeManifest: true,
      path: deployPath,
      prettyPrint: true
    })
  ];

  plugins.push(new webpack.NamedModulesPlugin());
  plugins.push(new webpack.optimize.ModuleConcatenationPlugin());

  return plugins;
}

module.exports = {
  context: root,

  mode: prod ? 'production' : 'development',
  devtool: false, // rely on SourceMapDevToolPlugin
  cache: dev,

  stats: {
    chunks: false,
    chunkModules: false,
    entrypoints: false,
    modules: false,
    version: false,
    hash: false,
    timings: false,
  },

  entry: {
    app: './core-app/bootstrap.ts',
    embed: './core-embed/bootstrap.ts'
  },

  watchOptions: {
    aggregateTimeout: 300,
    ignored: /node_modules/
  },

  resolve: {
    extensions: ['.ts', '.tsx', '.js', '.jsx'],
    alias: {
      handsontable: 'handsontable/dist/handsontable.full.js'
    }
  },

  output: {
    filename: outputFilename,
    path: path.resolve(__dirname, '../public/assets'),
  },

  optimization: {
    runtimeChunk: {
      name: "manifest",
    },
    splitChunks: {
      cacheGroups: {
        // default: false,
        commons: {
          test: /[\\/]node_modules[\\/]/,
          name: "vendor",
          chunks: "initial",
          enforce: true
        }
      },
    },
  },

  externals: {
    nvd3: 'window.nv',
    highcharts: "window.Highcharts",
    angular: 'window.angular',
    jquery: 'window.jQuery',
    lodash: 'window._',
    moment: 'window.moment',
    ace: 'window.ace',
    d3: 'window.d3',
    chiliPiper: 'window.ChiliPiper',
  },

  plugins: getPlugins(),

  module: {
    rules: [
      {
        test: /\.ts(x?)$/,
        exclude: [/\.(spec|e2e)\.ts$/],
        use: [{
          loader: 'ts-loader',
          options: {
            experimentalWatchApi: true // seems to work ok atm, but performance improvement varies
          }
        }]
      }, {
        test: /\.(html|css)$/,
        loader: 'raw-loader',
        exclude: /\.async\.(html|css)$/
      }, {
        test: /\.svg$/,
        loader: 'url-loader?' + JSON.stringify({
          name: '[name]_[hash]'
        })
      }
    ]
  }
};
```

**Comment by michael-ciniawsky** [comment]
```diff
- const outputFilename = dev || ci ? '[name].bundle.js' : '[name].[chunkhash].bundle.js';
+ const outputFilename = dev || ci ? '[name].bundle.js' : '[name].[contenthash].bundle.js';
```

TLDR: Use `[contenthash]` instead of `[chunkhash]` to ensure that every asset (output file) has a stable hash based on it's contents. 

`[chunkhash]` isn't very 'stable'  as it can change when e.g a module of a chunk is moved for some reason etc etc (It's always hard to tell why and will always be based on how `[chunkhash]` works (in terms of browser caching)). For predictable long term (browser) caching the contents of the asset are the only reliable hashing source here since browsers cache files based on `Map { url => content }` relations without any knowledge of webpack chunks or the like...

**Comment by michael-ciniawsky** [comment]
Also don't forget to set the `chunkFilename` for Code Splitting (Async Chunks)

```js
const output = {
  filename: dev || ci ? '[name].bundle.js' : '[name].[contenthash].bundle.js',
  chunkFilename: dev || ci ? '[name].chunk.js' : '[name].[contenthash].chunk.js'
}
```

**Comment by mikeyoon** [comment]
I changed it to contenthash, but I still get different hashes and they have the same content. I even took out the sourcemap output to make sure they are exactly the same.

**Comment by viktorlarsson** [comment]
Experiencing the same issue, did you resolve it?

**Comment by evilebottnawi** [comment]
@mikeyoon can you create minimum reproducible test repo?

**Comment by mikeyoon** [comment]
@evilebottnawi I tried, but wasn't able to narrow it down because I don't know what causes the issue or how the logic works.

**Comment by evilebottnawi** [comment]
@mikeyoon can we close issue? if problem still appear we can reopen

**Comment by mikeyoon** [comment]
The issue has already been closed earlier, although it definitely still happens to me and I have chosen not to upgrade to Webpack 4 because of it. I don't really care too much at this point as v3 still works for us.

**Comment by MQuy** [comment]
it happens for me too and I did use `contenthash` but they are changed per my build 😢

**Comment by jacobwindsor** [comment]
I also have this issue with chunkhash but I am seeing a pattern:

- Only the chunkhash of one "vendor" splitChunk is affected
- Only the hash is changed and not the file
  - I am using sourcemaps so the file does change but only on the line referencing the sourcemap
- The hash remains consistent through builds started in the same bash
  - I.e. if I run the build several times from git bash it's all good. Only if I run the build from VS integrated terminal and then from git bash is there a difference.

**Comment by ArtemZag** [comment]
Probably the hash is different because it based on module IDs that are generated during the build.

**Comment by mikeyoon** [comment]
I have HashedModuleIdsPlugin enabled, so there is probably some other factor.

**Comment by ArtemZag** [comment]
@mikeyoon , try also to enable NamedChunksPlugin

**Comment by mikeyoon** [comment]
Didn't work with [contenthash] or [chunkhash] unfortunately. There's definitely something with our codebase that's triggering the issue. Not all of our chunks have inconsistent hashes.

**Comment by jakub-g** [comment]
FYI
We had the same issue in our builds (though with webpack 3) when using `chunkhash` and I tracked it down to `script-loader`.
https://github.com/webpack-contrib/script-loader/issues/56

I think plugins and loaders (`raw-loader`) are most likely to contribute to the issue.
Right now I'm in the process of migration to webpack 4 and `contenthash`, will report my findings in the coming weeks.

**Comment by jakub-g** [comment]
Update: from what I've noticed in the recent weeks, the issue actually happens also for some new chunks in our project that are not loaded via `script-loader`. **The issue is not script-loader** then (probably).

On my machine, for a given git revision, `contenthash`es are deterministic, but when running a different git revision, some output chunks are often identical, but `contenthash` is different than a few days before.
 
It seems that the source of this is some internal webpack global build state, and that `contenthash` is not a reliable tool for the job, as of webpack4. I will try webpack5 in the coming weeks to see if it improves anything. If not, an alternative would be some hashing tool independent of webpack (after-build), as `contenthash` is also not taking into account the minification process (https://github.com/webpack-contrib/terser-webpack-plugin/issues/18#issuecomment-448622985), and I suspect many other webpack plugins/loaders may have the same issue.

**Comment by onlybye** [comment]
> Didn't work with [contenthash] or [chunkhash] unfortunately. There's definitely something with our codebase that's triggering the issue. Not all of our chunks have inconsistent hashes.

@mikeyoon have you solved  this problem ?

**Comment by evilebottnawi** [comment]
Please create minimum reproducible test repo

**Comment by mikeyoon** [comment]
I have not looked at the issue since I posted about it last time. I also haven't had a chance to try out the latest major release either. We're still just using v3.x for now.

**Comment by 779g** [comment]
Having the same issue on webpack 4, both with contenthash and chunkhash. If I find some time I will make a minimum repo. Shall the ticket be reopened ?

**Comment by halfnibble** [comment]
We are seeing this problem too. Unfortunately, it's inconsistent and the codebase is private. Could it have something to do with running 10+ webpack commands simultaneously in different threads? (We are building a mono repo managed by rush.

My assigned task is to determine why this is inconsistent, and to fix it, so any help you can provide would be much appreciated. Thanks.

**Comment by vcraescu** [comment]
Same issue here. We do async imports and because of this we can't run the app on multiple machines because the code is built separately on each machine so generated files names are not the same on all the machines.

**Comment by halfnibble** [comment]
I should have come back and posted an update.

Short answer, Webpack's hashing algorithm is deterministic. But the content being sent to it may not be. I discovered a whole slew of issues:
1. Some npm packages create absolute paths in postInstall scripts.
2. Some packages' error messages get sent to the hashing algorithm containing an absolute path.
3. And we had one webpack loader generating content containing absolute paths.

I found the easiest way to debug this issue was to override the webpack hashing algorithm and add a little script to detect for my local absolute path. Here is approximately what I used:
https://gist.github.com/halfnibble/c9f0c3b8d61e7002c799b5d24ae05643

And then in webpack config:
```typescript
{
  ...
  output: {
    hashFunction: HashFunction,
    ...
  }
}
```

**Comment by evilebottnawi** [comment]
@halfnibble can you create minimum reproducible test repo, maybe we can fix some cases on webpack side

**Comment by halfnibble** [comment]
@evilebottnawi As far as I can tell, there is absolutely nothing wrong with webpack's hashing algorithm. The problem lies in certain npm packages and webpack loaders. None of the problems I encountered were in code from the webpack organization. 

I don't know if it would be worthwhile to create a content hash debug feature like the one we developed to pin point the packages and loaders causing problems.

**Comment by evilebottnawi** [comment]
@halfnibble thanks for feedback, yep, some loaders/plugins emit assets with absolute path, better open issues in repos with loaders/plugins

**Comment by jakub-g** [comment]
@halfnibble re: https://github.com/webpack/webpack/issues/7179#issuecomment-547557053 

fantastic comment, thanks to it I revisited the topic and found that it was indeed `script-loader` that was breaking hashing in our project (due to the usage of `path.join` and `require.resolve` which both output absolute paths). I migrated away from it to `raw-loader` instead, and submitted the PR to deprecate `script-loader`.

The info about not using absolute paths is documented in "writing a loader" doc:

https://webpack.js.org/contribute/writing-a-loader/#absolute-paths
